### PR TITLE
Find all to find by object

### PR DESCRIPTION
**Pull request by andreicek**
The feat of this pull request is to add the ability to find by a plain object.

```javascript
store.findAll('developers', {age: '24', name: 'andrei'});
```

Please don't merge just yet.

### CODE REVIEW

**Comment by andreicek** [comment]
@DarkoKukovec ready for review.

File: src/Collection.ts [hunk 1]
```diff
@@ -182,10 +182,16 @@ export class Collection implements ICollection {
...
    * @memberOf Collection
    */
-  public find<T extends IModel>(type: IType, id?: string|number): T {
-    return id
-      ? ((this.__modelHash[type] && this.__modelHash[type][id]) || null)
-      : (this.__data.find((item) => getType(item) === type) as T) || null;
+  public find<T extends IModel>(type: IType, id?: string|number): T;
+  public find<T extends IModel>(type: IType, searchObject?: object, searchParams?: object): T;
+  public find<T extends IModel>(type: IType, searchPredicate?: string|number|object, searchParams?: object): T {
+    if (!searchPredicate || typeof searchPredicate !== 'object') {
```
Comment (andreicek) [inline src/Collection.ts]: there has to be a better way of doing this...

Comment (DarkoKukovec) [reply src/Collection.ts]: Not sure :/

File: src/Collection.ts [hunk 2]
```diff
@@ -182,10 +182,16 @@ export class Collection implements ICollection {
...
+      return searchPredicate
+        ? ((this.__modelHash[type] && this.__modelHash[type][searchPredicate as string|number]) || null)
+        : (this.__data.find((item) => getType(item) === type) as T) || null;
+    }
+
+    return (this.__data.find((item: T) => item.isEqual(searchPredicate as object, searchParams)) as T) || null;
```
Comment (DarkoKukovec) [inline src/Collection.ts]: Is this `as object` needed, since you have the if up there?

File: src/Model.ts [hunk 3]
```diff
@@ -346,6 +348,31 @@ export class Model implements IModel {
...
   /**
+   * Compares a regular object to a model.
+   *
+   * @param {object} comparingObject Object that will be compared to the model
+   * @param {IModelIsEqualParams} params Options for comparison
+   * @returns {boolean} Is the object equal to the model
+   * @memberof Model
+   */
+  public isEqual(comparingObject: object, params?: IModelIsEqualParams): boolean {
+    if (!params) {
```
Comment (DarkoKukovec) [inline src/Model.ts]: 1. Don't change arguments - nothing good can happen...
2. Why not use default values and set it as required (not nullable)?

File: src/Model.ts [hunk 4]
```diff
@@ -346,6 +348,31 @@ export class Model implements IModel {
...
+      params = {
+        ignoreId: true,
+      };
+    }
+
+    const runParams: IModelIsEqualParams = assign({}, params, {
+      ommitPaths: params.ommitPaths || ['__type__'],
+    });
+    if (runParams.ignoreId) {
+      runParams.ommitPaths.push('id');
```
Comment (DarkoKukovec) [inline src/Model.ts]: This is not quite correct. You should use `this.static.idAttribute`.

File: src/Model.ts [hunk 5]
```diff
@@ -346,6 +348,31 @@ export class Model implements IModel {
...
+   */
+  public isEqual(comparingObject: object, params?: IModelIsEqualParams): boolean {
+    if (!params) {
+      params = {
+        ignoreId: true,
+      };
+    }
+
+    const runParams: IModelIsEqualParams = assign({}, params, {
+      ommitPaths: params.ommitPaths || ['__type__'],
```
Comment (DarkoKukovec) [inline src/Model.ts]: Use the `TYPE_PROP` constant.

File: src/utils.ts [hunk 6]
```diff
@@ -80,3 +80,29 @@ export function assign(target: object, ...args: Array<object>) {
...
+export function omit(target: object, paths: Array<string>) {
+  if (!paths.length) {
+    return target;
+  }
+
+  return Object
+    .keys(target)
+    .filter((key) => paths.indexOf(key) < 0)
+    .reduce((acc, val) => {
+      if (target.hasOwnProperty(val)) {
```
Comment (DarkoKukovec) [inline src/utils.ts]: Why not use ternary operator?

File: src/Collection.ts (hunk 2 above)
Comment (andreicek) [reply src/Collection.ts]: unfortunately yes.

Comment (DarkoKukovec) [reply src/Collection.ts]: This doesn't seem right... Can you invert the if `if (typeof searchPredicate === 'object')`?

File: src/Model.ts [hunk 7]
```diff
@@ -346,6 +348,25 @@ export class Model implements IModel {
...
+   * Compares a regular object to a model.
+   *
+   * @param {object} comparingObject Object that will be compared to the model
+   * @param {IModelIsEqualParams} params Options for comparison
+   * @returns {boolean} Is the object equal to the model
+   * @memberof Model
+   */
+  public isEqual(comparingObject: object, params: IModelIsEqualParams = {ignoreId: true}): boolean {
+    const runParams: IModelIsEqualParams = assign({}, params, {
+      ommitPaths: params.ommitPaths || [TYPE_PROP],
```
Comment (DarkoKukovec) [inline src/Model.ts]: Actually, I was wrong... `this.static.typeAttribute` :)

File: src/Model.ts [hunk 8]
```diff
@@ -346,6 +348,25 @@ export class Model implements IModel {
...
+    });
+    if (runParams.ignoreId) {
+      runParams.ommitPaths.push(this.static.idAttribute);
+    }
+    const nativeObject = omit(this.toJS(), runParams.ommitPaths);
```
Comment (DarkoKukovec) [inline src/Model.ts]: On second throught... Why even create a whole `runParams` object if you only need an array?

File: src/utils.ts [hunk 9]
```diff
@@ -80,3 +80,28 @@ export function assign(target: object, ...args: Array<object>) {
...
+export function omit(target: object, paths: Array<string>) {
+  if (!paths.length) {
+    return target;
+  }
+
+  return Object
+    .keys(target)
+    .filter((key) => paths.indexOf(key) < 0)
+    .reduce((acc, val) => (
+      target.hasOwnProperty(val)
```
Comment (DarkoKukovec) [inline src/utils.ts]: Is this repeated use of `assign` fast enough? Wouldn't it be faster to create a clone of the object once, and then delete the omitted properties?

**Pull request merged**